File: fn/src/main/java/org/sqlbuilder/fn/joins/FE_FEGroupRealization.java
package org.sqlbuilder.fn.joins;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.common.RequiresIdFrom;
import org.sqlbuilder.fn.objects.FE;
import org.sqlbuilder.fn.objects.FEGroupRealization;
import org.sqlbuilder.fn.types.FeType;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

import edu.berkeley.icsi.framenet.FEValenceType;

public class FE_FEGroupRealization extends Pair<String, FEGroupRealization> implements Insertable
{
	public static final Set<FE_FEGroupRealization> SET = new HashSet<>();

	// C O N S T R U C T O R

	@SuppressWarnings("UnusedReturnValue")
	public static FE_FEGroupRealization make(final FEValenceType fe, final FEGroupRealization fegr)
	{
		var fr = new FE_FEGroupRealization(fe.getName(), fegr);
		SET.add(fr);
		return fr;
	}

	private FE_FEGroupRealization(final String feName, final FEGroupRealization fegr)
	{
		super(feName, fegr);
	}

	// A C C E S S

	String getFEName()
	{
		return first;
	}

	String getFENames()
	{
		return second.getFENames();
	}

	// O R D E R

	public static final Comparator<FE_FEGroupRealization> COMPARATOR = Comparator.comparing(FE_FEGroupRealization::getFEName).thenComparing(FE_FEGroupRealization::getFENames);

	// I N S E R T

	@RequiresIdFrom(type = FEGroupRealization.class)
	@Override
	public String dataRow()
	{
		// fegrid,feid,fetypeid
		int fetypeid = FeType.getIntId(first);
		var key = new Pair<>(fetypeid, second.getFrameID());
		var feid = FE.BY_FETYPEID_AND_FRAMEID.get(key).getID();

		return String.format("%s,%s,%s", //
				second.getSqlId(), //
				feid, //
				fetypeid);
	}

	@Override
	public String comment()
	{
		return String.format("fe=%s,fes={%s},luid=%d,frid=%d", first, getFENames(), second.getLuID(), second.getFrameID());
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[FE-FEGR fe=%s fegr={%s}]", first, second.getFENames());
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/joins/FE_FEGroupRealization.kt
package org.sqlbuilder.fn.joins

import edu.berkeley.icsi.framenet.FEValenceType
import org.sqlbuilder.annotations.RequiresIdFrom
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.fn.objects.FE
import org.sqlbuilder.fn.objects.FEGroupRealization
import org.sqlbuilder.fn.types.FeType.getIntId
import java.util.*

data class FE_FEGroupRealization(
    val feName: String,
    val fegr: FEGroupRealization,
) : Insertable {

    val fENames: String
        get() = fegr.fENames

    // I N S E R T

    @RequiresIdFrom(type = FEGroupRealization::class)
    override fun dataRow(): String {
        val fetypeid: Int = getIntId(feName)!!
        val key = Pair(fetypeid, fegr.frameID)
        val feid = FE.BY_FETYPEID_AND_FRAMEID!![key]!!.iD
        return "${fegr.getSqlId()},$feid,$fetypeid"
    }

    override fun comment(): String {
        return "fe=$feName,fes={$fENames},luid=${fegr.luID},frid=${fegr.frameID}"
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as FE_FEGroupRealization
        return feName == that.feName && fegr == that.fegr
    }

    override fun hashCode(): Int {
        return Objects.hash(feName, fegr)
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[FE-FEGR fe=$feName fegr={$fegr}]"
    }

    companion object {

        @JvmField
        val COMPARATOR: Comparator<FE_FEGroupRealization> = Comparator
            .comparing<FE_FEGroupRealization, String> { it.feName }
            .thenComparing<String> { it.fENames }

        @JvmField
        val SET = HashSet<FE_FEGroupRealization>()

        fun make(fe: FEValenceType, fegr: FEGroupRealization): FE_FEGroupRealization {
            val fr = FE_FEGroupRealization(fe.getName(), fegr)
            SET.add(fr)
            return fr
        }
    }
}
